core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/worker/WorkflowWorker.scala

```diff
@@ -0,0 +1,289 @@
+  IOperatorExecutor,
+  ISourceOperatorExecutor,
+  ITupleSinkOperatorExecutor
+}
+import edu.uci.ics.amber.error.WorkflowRuntimeError
+
+import scala.annotation.elidable
+import scala.annotation.elidable.INFO
+import scala.collection.mutable
+import scala.concurrent.ExecutionContext
+import scala.concurrent.duration._
+
+object WorkflowWorker {
+  def props(id: ActorVirtualIdentity, op: IOperatorExecutor): Props =
+    Props(new WorkflowWorker(id, op))
+}
+
+class WorkflowWorker(identifier: ActorVirtualIdentity, operator: IOperatorExecutor)
+    extends WorkflowActor(identifier) {
+  implicit val ec: ExecutionContext = context.dispatcher
+  implicit val timeout: Timeout = 5.seconds
+
+  lazy val pauseManager: PauseManager = wire[PauseManager]
+  lazy val dataProcessor: DataProcessor = wire[DataProcessor]
+  lazy val dataInputPort: DataInputPort = wire[DataInputPort]
+  lazy val dataOutputPort: DataOutputPort = wire[DataOutputPort]
+  lazy val batchProducer: TupleToBatchConverter = wire[TupleToBatchConverter]
+  lazy val tupleProducer: BatchToTupleConverter = wire[BatchToTupleConverter]
+  lazy val promiseManager: WorkerPromiseManager = wire[WorkerPromiseManager]
+  lazy val workerStateManager: WorkerStateManager = wire[WorkerStateManager]
+
+  val receivedFaultedTupleIds: mutable.HashSet[Long] = new mutable.HashSet[Long]()
+  var isCompleted = false
+  @elidable(INFO) var startTime = 0L
+
+  def onSkipTuple(faultedTuple: FaultedTuple): Unit = {
+    if (faultedTuple.isInput) {
+      dataProcessor.setCurrentTuple(null)
+    } else {
+      // if it's output tuple, it will be ignored
+    }
+  }
+
+  def onResumeTuple(faultedTuple: FaultedTuple): Unit = {
+    if (!faultedTuple.isInput) {
+      batchProducer.passTupleToDownstream(faultedTuple.tuple)
+    } else {
+      dataProcessor.prependElement(InputTuple(faultedTuple.tuple))
+    }
+  }
+
+  def onModifyTuple(faultedTuple: FaultedTuple): Unit = {
+    if (!faultedTuple.isInput) {
+      batchProducer.passTupleToDownstream(faultedTuple.tuple)
+    } else {
+      dataProcessor.prependElement(InputTuple(faultedTuple.tuple))
+    }
+  }
+
+  def getInputRowCount(): Long = {
+    val (inputCount, _) = dataProcessor.collectStatistics()
+    inputCount
+  }
+
+  def getOutputRowCount(): Long = {
+    val (_, outputCount) = dataProcessor.collectStatistics()
+    outputCount
+  }
+
+  def getResultTuples(): mutable.MutableList[ITuple] = {
+    this.operator match {
+      case processor: ITupleSinkOperatorExecutor =>
+        mutable.MutableList(processor.getResultTuples(): _*)
+      case _ =>
+        mutable.MutableList()
+    }
+  }
+
+  override def receive: Receive = receiveAndProcessMessages
+
+  def receiveAndProcessMessages: Receive = {
+    routeActorRefRelatedMessages orElse
+      processControlMessages orElse
+      oldControlMessageHandler orElse
+      receiveDataMessages orElse {
+      case other =>
+        logger.info(s"unhandled message: $other")
```

Comment:
I suggest we log an error in these cases. Also I think we can go a bit extreme and actually end the workflow running because of such errors which suggest something wrong going on with the workflow. What do you think?

Reply:
Also use the WorkflowLogger :)

Reply:
I feel the WorkflowLogger API is hard to use especially when I need to log errors. We can overload logError to let it only take a string so that we don't have to create WorkflowRuntimeError for every call